# Tutorial 02

## 0 Review()

In [ ]:
# 숫자
# integer(정수) : index, slicing, counting
# float(실수) : 연산

# 문자
# 'datum' :  문자와 문자의 사칙연산,  Indexing/ Slicing,  Counting....
# data  :  매개변수

# [ list ]   :  integer index 값이 숨어있다. (0 ~ -1)

# { key : value }

<p>출처 : http://pbpython.com/pandas-list-dict.html</p>
</br><img src = "http://pbpython.com/images/pandas-dataframe-shadow.png" align="left">

In [111]:
### 앞에서 사용된 기본함수들
# len()
# type()
# int()
# float()
# str()
# list()
# dict()

In [6]:
lists = [1, 2, 3, '회계', '재무', 2017, ['08-15', '22:00:30'], ['Python', 'Jupyter'], 'apple']

## 3 Function (기본함수/모듈, 외부함수/ 모듈)
반복문, 조건문

### 01. 반복문
for - list / dict 객체들을 순차적 반복 처리

In [147]:
# ':' 부분집합으로 아래의 내용은 위 함수의 조건 내에서만 실행
# indent (안으로 약간 들여쓰다) - 부분집합에 해당을 인식

for list_item in lists :   
    print(list_item)       

1
2
3
회계
재무
2017
['08-15', '22:00:30']
['Python', 'Jupyter']


In [151]:
for key in dictionary:
    print(key)

005390
035420


In [155]:
for key , value in dictionary.items() :
    print(key, value)

005390 삼성전자
035420 네이버


In [156]:
# enumerate()
for index, list_item in enumerate(lists) :
    print(index , list_item)

0 1
1 2
2 3
3 회계
4 재무
5 2017
6 ['08-15', '22:00:30']
7 ['Python', 'Jupyter']


In [18]:
# range( 종료값 )  
# 종료값 - Counting : integer
for i in range(4): print(i)

0
1
2
3


In [20]:
# range( 시작, 종료, step)
# 시작, 종료값 , step - Counting : integer
for i in range(0,100,20): print(i)

0
20
40
60
80


In [23]:
for i in range(100,10,-20): print(i)

100
80
60
40
20


In [25]:
# range( 시작, 종료, step)
# step : float (실수는 불가능하다!!)
# 기본 법칙에서 위배된 실행은 불가능하다
for i in range(0,1,0.2): print(i)

TypeError: 'float' object cannot be interpreted as an integer

In [26]:
# 기본 규칙을 벗어나려면....
# Numpy 등 외부모듈을 사용해야 한다.
import numpy as np
np.arange(0, 1, 0.1)

array([ 0. ,  0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9])

### 02. 판단문 - Boolean 연산자 (컴퓨터 용어)
if - [True, False, True, True ......] 딱 2가지로만 출력

Masking (데이터 선별기능)

In [159]:
for out in lists:
    if type(out) == int:
        print(out)

1
2
3
2017


In [157]:
for out in lists:
    if type(out) == str:
        print(out)

회계
재무


In [158]:
for out in lists:
    if type(out) == list:
        print(out)

['08-15', '22:00:30']
['Python', 'Jupyter']


In [160]:
for index, out in enumerate(lists):
    if type(out) == list:
        print(index, out)

6 ['08-15', '22:00:30']
7 ['Python', 'Jupyter']


### 03. [  실행함수            반복문     조건문] --  [ list 함수]
 1. [ 동사      명사와 형용사 ] 
 2. 영어의 '명령문' 형식을 [ list ] 내부에 차용
 3. 단 1줄에.....  [최종실행함수   반복문 조건문]  3개의 함수의 실행결과를 list로 일괄출력

반복문 / 조건문 충족 --> 최종 실행함수 --> 순차적 list 로 출력

In [7]:
for index, out in enumerate(lists):
    if type(out) == list:
        print(index, out)

6 ['08-15', '22:00:30']
7 ['Python', 'Jupyter']


In [11]:
[(index, out)     for index, out in enumerate(lists)]

[(0, 1),
 (1, 2),
 (2, 3),
 (3, '회계'),
 (4, '재무'),
 (5, 2017),
 (6, ['08-15', '22:00:30']),
 (7, ['Python', 'Jupyter']),
 (8, 'apple')]

In [13]:
[(index, out)     for index, out in enumerate(lists)  if type(out) == str]

[(3, '회계'), (4, '재무'), (8, 'apple')]

## QUIZ 01

range() 함수를 사용해서 2010 ~ 2017 까지 숫자 중 2 단계씩의 숫자를 출력하라

    tip)  range( start, end, step)

## QUIZ 02

range() 함수를 사용해서 2017 ~ 1990 까지 숫자 중 4의 배수만 list로 출력하라

    tip) 2014 % 4 == 0 : 4로 나눌 떄 나머지가 없을 것

## 4 연결기호
그냥 쓰다보면, 나중에 큰 코 다치는 주요한 부분

In [ ]:
## 01 =
#  (실행) 데이터 입력 

a = ['2017-01-01']

In [28]:
## 02 ==
#  (판단) 조건문 
[i         for i in range(10)       if i % 2 ==0]

[0, 2, 4, 6, 8]

In [ ]:
## 01 ' ' (단 따옴표) ,  " " (쌍 따옴표)  
#  ex) '제주도'는 '한반도'의 부속도서로써...
#  이것이 문자열이다

"""   Hello this is web browser text book in jupyter notebook,
you can edit in web brower , so could be ues in server version """

In [161]:
## 02 . 
# ex) 3.141592....
# 종속-1 : .메소드함수()
dictionary.items()
lists.append('apple')
lists

[1,
 2,
 3,
 '회계',
 '재무',
 2017,
 ['08-15', '22:00:30'],
 ['Python', 'Jupyter'],
 'apple']

In [161]:
## 03 ,     
# ex) 삼성, Lg, SK, Kt...
# 대등관게 (비중이 같은 관계) 
for k ,v  in dictionary.items():
    print('Code : ', k , '  name : ',v)

[1,
 2,
 3,
 '회계',
 '재무',
 2017,
 ['08-15', '22:00:30'],
 ['Python', 'Jupyter'],
 'apple']

In [3]:
## 04 : 
# ex) 파이썬 : 이번 학습의 주요 아이템이다.
# 내부를 Detail하게 서술 1) 코딩내용은 함수에 종속되어서 실행
#                    2) .메소드함수() 와 비교해서 사용자 편집성이 더 강하다
#                    3) indent 가 꼭 따라온다/
#                    4) 단 1줄로 처리 가능시 바로 뒤에 연달아서 입력 가능

for out in lists: print(out)

1
2
3
회계
재무
2017
['08-15', '22:00:30']
['Python', 'Jupyter']
apple


## QUIZ 01
    [i         for i . v  in enumerate(range(10))       if i % 2 = 0 ]

    를 실행하면 Syntax Error 가 발생한다. 오류 2개가 무엇인지 설명하시오

In [32]:
[i         for i . v  in enumerate(range(10))       if i % 2 = 0 ]

SyntaxError: invalid syntax (<ipython-input-32-cdb35ef778e8>, line 1)

## 5 .formating   in String
엑셀의 사용자함수 - 단어의 순서 및 형식을 변경/ 정의가능

%d, %s  -->  %매개변수 or %21

### 01 %d  : digit 
문자열 속 숫자(integer) 매개변수 아라비아 숫자

In [38]:
# %d  digit : 아라비아 숫자
# string 속에서 %d 임시 매개변수를 지정

'I am %d years olds' %18

'I am 18 years olds'

In [43]:
# %d  digit : 아라비아 숫자
# string 속에서 %d 임시 매개변수를 지정

'I am %d years olds' %3.1415

'I am 3 years olds'

In [54]:
# %f  float : 실수

age = 21.3
'I am %f years olds' %age

'I am 21.300000 years olds'

In [35]:
age = '21'
'I am %d years olds' %age

TypeError: %d format: a number is required, not str

### 02  %s  : string 
문자열 속 문자 매개변수

In [50]:
# %s  String
age = '21'
'I am %s years olds' %age

'I am 21 years olds'

In [45]:
# %s  String
'He is a famous %s in Korea' %"Accountant"

'He is a famous Accountant in Korea'

### 03  formating 변수가  여럿 사용하는 경우 
문자열 속 문자 매개변수

In [46]:
number = 10
day = '30'
'I ate %d Apples, so I sick for %s days' %(number, day)   # %(0번 순서, 1번 순서)

'I ate 10 Apples, so I sick for 30 days'

In [47]:
'I ate %d Apples, so I sick for %s days' %(2, 'two')   # %(0번 순서, 1번 순서)

'I ate 2 Apples, so I sick for two days'

In [49]:
# %% 
'This program has %d%% error rate' %(9)

'This program has 9% error rate'

In [51]:
# %% 
'This program has %f%% error rate' %(0.156)

'This program has 0.156000% error rate'

In [55]:
# %c : character 문자 1개
# %s : String 문자열
# %d : integer 정수
# %f : floating point 부동소수

## QUIZ 01
    "2017년 대한민국 상위 회계사 0.111%가 우리 수업을 수강하고 있습니다" 를
    %f 0.111% , %d 2017 , %s '회계사' 를 활용해서 만들어보세요

In [ ]:
"% 년 대한민국 상위 %  %% 가 우리 수업을 수강하고 계십니다"

### 04 자릿수 조정하기
default : 오른쪽 정렬

    %c : character 문자 1개
    %s : String 문자열
    %d : integer 정수
    %f : floating point 부동소수

In [26]:
## String

#%6  s  (전체 입력공간의 최소크기를 지정)
"hi my friend.... %6s" %"Hello come to the real world"

'hi my friend.... Hello come to the real world'

In [27]:
#%-6  s  (전체 입력공간의 최소크기를 지정)
# -6 : 왼쪽부터 정렬
"Oh!! my girl.... %-6s jane" %"hello"

'Oh!! my girl.... hello  jane'

In [28]:
## Digit

#%6 (전체 입력공간 크기 값)
"your ID number is  %6d" %1020

'your ID number is    1020'

In [29]:
## float

#%2f (입력은 가능하나 인식은 못 함)  # 코딩시 실수하는 부분!!
"Pi * 100 is   %2f" %3123.141592

'Pi * 100 is   3123.141592'

In [30]:
#%.2f (소숫점 자리 수)
"Pi * 100 is   %0.2f" %3123.141592

'Pi * 100 is   3123.14'

In [31]:
#%.2f (소숫점 자리 수)
"Pi is   %.5f" %3.123141592

'Pi is   3.12314'

## QUIZ 01
    price = ['100원' , '20,000원' , '1,000,000원']
    를 총 10개칸의 오른쪽 정렬로 일괄 출력을 하자

In [77]:
price = ['100원' , '20,000원' , '1,000,000원']
for i in price:
    print('%10s'%i)

      100원
   20,000원
1,000,000원


### 05 기타 string 관련 메소드 들
아.. 이런게있구나 알아두는 정도

실무 - Web Crawling/ pdf 문서의 오타/ 파일의 이름 변경등.. 다양한 전처리.. import re 를 더 많이 쓴다 ('정규표현식'의 문법에 익숙하므로)

    .count()
    .find()
    .index()
    .join()

    .strip()   # 모든 공백제거
    .lstrip()  # 왼쪽 공백만 제거
    .rstrip()  # 오른쪽 공백만 제거

    .replace()
    .strip()

In [1]:
a = 'Accoutant'

In [3]:
# .count()
a.count('A')

1

In [4]:
a.count('c')

2

In [5]:
# .index()
a.index('t')

5

In [7]:
a.index('z')

ValueError: substring not found

In [8]:
# .join()
a.join('Korea')

'KAccoutantoAccoutantrAccoutanteAccoutanta'

In [9]:
'Korea' + a

'KoreaAccoutant'

In [11]:
# replace()
b = ' I am a Japan Accountant'
b.replace('Japan', 'South Korea')

' I am a South Korea Accountant'

In [12]:
# split()   # 기본은 '공백'을 기준으로 나눈다
b.split()

['I', 'am', 'a', 'Japanease', 'Accountant']

In [13]:
c = '이번\t결산월\t회계작업에\t투입\t가능한\t인원은\t2명\t에\t불과하다'
c.split('\t')

['이번', '결산월', '회계작업에', '투입', '가능한', '인원은', '2명', '에', '불과하다']

## Quiz 
    import glob
    filename = glob.glob('./*.*') 
    을 for 반복문을 활용해서 폴더 , 파일이름 , 확장자를 구별해서 화면에 출력하라

In [19]:
import glob
filelist = glob.glob('./*.*')

In [20]:
for file in filelist:
    print(file.split('.'))

['', '/daum', 'csv']
['', '/04', 'Wavelet analysis', 'ipynb']
['', '/02', 'functions_method', 'ipynb']
['', '/Tut_Finance01', 'ipynb']
['', '/03', 'pandas', 'ipynb']
['', '/sklearn01', 'intro', 'ipynb']
['', '/intro', 'py']
['', '/01', 'Introduction', 'ipynb']
['', '/05', 'Time_Series', 'ipynb']
['', '/Tut_Finance02', 'ipynb']


### 06 .format()
'index값' 과 '속성값' 2개를 {dict} 포맷을 활용해서 적용한다

In [23]:
# ' { index } '.format( tuple )  # 변하지 않을 값을 사용자가 입력한다, tuple의 활용 예.. 
"I {0} {1} apple".format('eat','apple') 

'I eat apple apple'

In [24]:
# {} 객체를 쓰면 반복적인 작업에도 유용하다
"I {0} {1} apple {0}".format('eat','apple') 

'I eat apple apple eat'

In [25]:
# { index : 0.2f } 2자리까지 출력
number = 3.141592
day = "three"
"I ate {0:0.2f} apples. so I was sick for {1} days".format(number, day)

'I ate 3.14 apples. so I was sick for three days'

In [1]:
# { index : % } 해당 float를 %로 변환해서 출력
number = 3.141592
day = "three"
"I ate {0:%} apples. so I was sick for {1} days".format(number, day)

'I ate 314.159200% apples. so I was sick for three days'

In [2]:
# { index : e } 해당 float를 10진수 포맷으로 변환해서 출력
number = 3.141592
day = "three"
"I ate {0:e} apples. so I was sick for {1} days".format(number, day)

'I ate 3.141592e+00 apples. so I was sick for three days'

In [32]:
# { index : <  왼쪽 정렬 자릿수 }
number = 3.141592
day = "three"
"I ate {0:<16} apples. so I was sick for {1} days".format(number, day)

'I ate 3.141592         apples. so I was sick for three days'

In [33]:
# { index : >  오른쪽 정렬 자릿수 }
number = 3.141592
day = "three"
"I ate {0:>16} apples. so I was sick for {1} days".format(number, day)

'I ate         3.141592 apples. so I was sick for three days'

In [34]:
# { index : ^  가운데 정렬 자릿수 }
number = 3.141592
day = "three"
"I ate {0:^16} apples. so I was sick for {1} days".format(number, day)

'I ate     3.141592     apples. so I was sick for three days'

In [35]:
# { index : 공백 채울 내용 ^ 자릿수 }
number = 3.141592
day = "three"
"I ate {0:-^16} apples. so I was sick for {1} days".format(number, day)

'I ate ----3.141592---- apples. so I was sick for three days'

## Quiz 1

    pi = 3.14159215
    .format 함수를 활용해서
    'Pi number is 3.14'로 출력하라

    hint>> "Pi number is {0:  }".format()

## Quiz 2

    ['+0.17%', '-0.25%', '-0.17%', '-2.49%', '+0.82%', '+0.83%']
    결과를 %를 삭제한 소숫점 5자리 실수로 변경해보자
    
    hint>> .replce('%', '') 

In [37]:
from intro import naver_short_table
table = naver_short_table('005930')
rates = list(table.등락률)
rates

['+0.17%', '-0.25%', '-0.17%', '-2.49%', '+0.82%', '+0.83%']

In [46]:
for rate in rates:
    rate = rate.replace('%', '')  # % 가 제외된 실수형태의 string
    rate = float(rate)/100        # string 을 float으로 변경 후 100을 나눈다
    print("%0.5f" % rate)         # float 결과를 5자릿수를 맞춰서 화면에 출력한다

0.00170
-0.00250
-0.00170
-0.02490
0.00820
0.00830


## 6  [ list ] 객체 가공 총정리
Python 의 자료관리의 Basic 기초, 다른 자료형 공부시 항상 [ list ] 에서 파생하며 공부 할 것!!

### 01 자료 살펴보기
indexing, slicing

In [65]:
a = [ 1 , 2 , 3 , [11, 12, 13]]  # 2차원 자료

In [66]:
# Slicing
a[2:]

[3, [11, 12, 13]]

In [67]:
a[-1]

[11, 12, 13]

In [68]:
a[-1][0] # slice in 2dim

11

### 02 Appending
자료 추가하기

In [69]:
a.append(b)
a

[1, 2, 3, [11, 12, 13], [10, 20, 30]]

In [70]:
a[0] = 100  # 데이터 교체
a

[100, 2, 3, [11, 12, 13], [10, 20, 30]]

In [71]:
# .insert( index , data )  
# 데이터를 index 위치에 추가하고, 
# index 뒤 데이터들을 1자리씩 뒤로 자동으로 옮겨준다
a.insert(0, [100, 200, 300])
a

[[100, 200, 300], 100, 2, 3, [11, 12, 13], [10, 20, 30]]

In [72]:
b = [10, 20, 30]
c = a + b
c

[[100, 200, 300], 100, 2, 3, [11, 12, 13], [10, 20, 30], 10, 20, 30]

### 03 Delete
자료 삭제하기

In [73]:
# .append() 연산직후, 결과를 되돌릴때, insert 로 늘어난 길이를 1 줄일떄
# .pop() 으로 날라가는 자료를 화면에 출력
a.pop()  

[10, 20, 30]

In [78]:
# slicing 문법을 활용해서 삭제하기
c[2:5] = []
c

[[100, 200, 300], 100, [10, 20, 30], 10, 20, 30]

In [79]:
# index 를 활용해서 삭제하기
# 빈 [ list ] 가 대신 입력되어서 깔끔하지 못하다
c[0] = []
c

[[], 100, [10, 20, 30], 10, 20, 30]

In [81]:
del c[0]
c

[100, [10, 20, 30], 10, 20, 30]

In [82]:
# .remove()  # value 값을 지정하면 해당하는 최초 index 값만 삭제한다
b.remove(30)  

[20, 30]

In [83]:
d = [10, 10, 20, 30, 50, 10, 60 ,80, 10]
d.remove(10)
d

[10, 20, 30, 50, 10, 60, 80, 10]

### 05 .sort()
기본설정 : 오름차순

In [89]:
a = [ i  for i in range(10,0,-1)]
a

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

In [90]:
a.sort()
a

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [93]:
# reverse 옵션을 활성화
a.sort(reverse=True)
a

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

In [95]:
# 알파벳 순서도 똑같이 정렬을 해준다
korea = ['Seoul','In-chon','Pusan', 'Daegu']
korea.sort()
korea

['Daegu', 'In-chon', 'Pusan', 'Seoul']

In [96]:
# 한글의 규칙도 알아서 적용 (시스템 상의 언어팩 데이터를 기초로 정렬)
korea = ['서울', '대전', '대구', '부산']
korea.sort()
korea

['대구', '대전', '부산', '서울']

In [97]:
# .reverse() : 있다는 건만 눈으로 확인, 실무에선 위 .sort() 메소드만 쓰자!
korea.reverse()
korea

['서울', '부산', '대전', '대구']

In [98]:
# [ list ] 데이터 활용하기
# 찾기   : [n1] indexing, [n1:n2] slicing
# 더하기  : .append(객체) , .insert(index주소, 객체) , + 
# 제거하기 : .remove(객체), .pop() , list[n1:n2] = [] , del list[index주소]
# 정렬 : .sort(reverse=False) , .reverse()